### DISCUSSION TITLE
Retaining Dynamic Content Post-Websocket Reconnection in NiceGUI: A Streamlit User's Perspective

### DISCUSSION THREAD

**Discussion by eshyun** [zauberzeug/nicegui#1100] [Ideas / Feature Requests]
Hello, I'm an enthusiast of Streamlit who has recently delved into NiceGUI. I'm thoroughly enjoying my transition, especially the way NiceGUI manages events and CSS - it’s a clear advantage over Streamlit. My exploration so far has left me with little desire to revert back to Streamlit. 

However, one distinct aspect I've noticed about Streamlit is that its webpage content remains persistent, irrespective of any alterations. On the other hand, NiceGUI, with its unique architecture, restores the page contents to their initial state by clearing all dynamic content.

Here's an example to illustrate my point:
```python
from nicegui import ui

@ui.page('/')
def home():
	ui.label('This is static')
	output = ui.label('This is dynamic')
	input = ui.input().on('keypress.enter', lambda: output.set_text(input.value))

ui.run()
```
In this code, if you enter any text in the input field and press enter, the `output` label will change. Now, the label `This is static` continues to be displayed even after a client reconnects. However, the dynamic part - `This is dynamic` - will reset to its default state after the client disconnects and reconnects (post-websocket timeout).

My dilemma lies in wanting the dynamic content to remain consistent, even after a websocket reconnection. This is particularly an issue while using mobile devices. Shifting between the mobile browser and other apps can often result in disconnecting the connection. Upon reopening the browser, I find that NiceGUI has cleared all dynamic content, which can be frustrating.

Given NiceGUI's current architecture, I understand that maintaining dynamic content could be challenging. However, I'm curious to know if there are others who feel the need for this feature in NiceGUI, or if this is simply a niche request. 

Regardless, I'm grateful for the exceptional package that NiceGUI is. I'm hoping that someone could possibly offer insights into this matter. Thanks!

**Comment by kleynjan** [comment]
Have you considered keeping persistent UI state in app.storage.user? I would expect that to "survive" websocket disconnect/reconnect.

**Comment by rodja** [comment]
Thanks for reaching out @eshyun. We understand that there are some scenarios where you want to retain the state of the system. There are several approaches:

1. Use the [auto-index page](https://nicegui.io/documentation#auto-index_page) which automatically retains the state and presents the same content to all connected clients (eg. a dialog opened by one client is also showing on all other clients).
2. As @kleynjan suggested you could use [binding with app.storage.user](https://nicegui.io/documentation/bindings#bind_to_storage) to keep specific state for each user -- even between server restarts.
3. Similar to 2. you can use `app.storage.general` to save state globally for all clients
4. Use global variables to hold the state like we did in the [chat app example](https://github.com/zauberzeug/nicegui/blob/7faadeecee14cbfaf7307d5c3b51f0a60a507522/examples/chat_app/main.py#L8).

This options show that the current NiceGUI architecture allows for much more flexibility as what you can do with Streamlit.

**Comment by eshyun** [comment]
Thank you for your input, @rodja and @kleynjan. I agree that there are various workarounds that can be utilized to address this particular problem. 

To illustrate, let's consider a chat application as an example. In this scenario, every time a client reconnects to the chat, the application re-renders all the messages. This poses a challenge because the client, post-reconnection, is not recognized as the original user from before the disconnection. As a result, the `user_id` can't be associated with the old one. Consequently, my messages appear as if they have been sent by another user.

While re-rendering simple messages might not pose a significant problem, the issue becomes more pronounced when it involves more complex tasks. For instance, if the process involves heavy computation or chart rendering, re-executing these tasks with each reconnection consumes excessive time and resources. This recurring process ultimately slows down the system, compromising its efficiency and user experience.

Therefore, it'd be great if I can find a solution that can maintain client-side UI (rendering) through reconnections, reducing the need for constant re-rendering of data and improving the application's overall performance.

**Reply by PranitModak** [reply]
I think you have to persist a user login along with id and then render it for the chat. because I am using authentication based login and rendering the chat message and it works perfectly for me.

Also I am using app.storage.user to store the user details and then fetch chat messages for the user.
So it works perfectly for me.

**Reply by rodja** [reply]
I want to second @PranitModak's response. 

>  every time a client reconnects to the chat, the application re-renders all the messages

Sure. Thats the nature of web-apps. The browser needs to render the content after connection. Thats also true for Streamlit. The question is, which content is send to the browser for re-rendering.

> This poses a challenge because the client, post-reconnection, is not recognized as the original user from before the disconnection

Why do you write that? `app.storage.user` recognizes the user. If you want to do it "by hand", you can use `app.storage.browser['id']` to get the unique id of a user.

> if the process involves heavy computation or chart rendering, re-executing these tasks with each reconnection consumes excessive time and resources

True. If your backend is not smart about caching the results of heavy computation, the content which is send to a client needs to be recomputed. A Streamlit-way of implicit caching can be archived by using the auto-index client, as I wrote above. By omitting the `@ui.page` decorator function, NiceGUI will serve all content to all clients.

**Comment by MohAid** [comment]
Hi, thanks for opening this discussion, I am really in the same position, I am new to nicegui I liked it but I stacked with this, I am working on a page that has a lot of entries the user has to enter and then click a button to save them (let's say something like a sales invoice), but on reloading they are all deleted and the user has to enter them from the beginning, what is the best advice you give to me for such a situation? (this situation could be wider than that but here is only a case example), or do you see nicegui is not suitable for alike types of websites?

**Reply by rodja** [reply]
Have a look at https://github.com/zauberzeug/nicegui/discussions/1100#discussioncomment-6310940

**Reply by MohAid** [reply]
Thanks for the help @rodja, but actually I've read all the comments here before I ask, so let me try to explain what confuses me:
Suppose that a user is trying to add two new invoices on two **tabs** of the page "/add_invoice"
how to bind the invoice details (table content) **depending on the opened tab**? 
I'm new to nicegui, am I missing something that had to be known to me?

**Reply by falkoschindler** [reply]
@MohAid If I understand correctly, you have a user connected to /add_invoice in two browser tabs. Both tabs show a bunch of inputs filled out be filled by the user. Now there are two requirements:

1. When the user reloads a tab, the input content should be restored.
2. Both tabs shouldn't interfere with each other, i.e. they can contain different content per tab.

I'm not sure how the server should know that the reloaded tab is supposed to have the same content like before. It could as well be a new tab with empty inputs, after the user closed the previous tab.

**Reply by MohAid** [reply]
Thanks @falkoschindler, you well understood the case, actually I expected this answer, this is the logic.
but why I asked? because this is not the user who reloads the page, nicegui built to do that frequently (after every connection loosing)! 
I wished to hear something solves that or stops this behavior, that's why I asked.

is it really impossible to achieve that in nicegui? I think most SaaS needs that!
I liked nicegui and I do not want to leave it as much as that possible.

**Reply by falkoschindler** [reply]
Do you know the `reconnect_timeout` in `ui.run` and `ui.page`? It allows you to specify how long NiceGUI waits for a client to reconnect. Maybe increasing it helps in your case? As long as the connection can be recovered within this interval, the input fields should keep their state.

**Reply by rodja** [reply]
> this is not the user who reloads the page, nicegui built to do that frequently (after every connection loosing)

Why are you seen so many connection drops? Are you accessing the server through a very unstable and bad internet connection? Then increasing `reconnect_timeout` could help. Also persisting the inputs similar to what @kleynjan did in https://github.com/zauberzeug/nicegui/discussions/1100#discussioncomment-7586923 is a good practice. But if the reloads happen due to cpu intensive processing you should read https://github.com/zauberzeug/nicegui/wiki/FAQs#how-to-not-get-the-reconnect-popup-or-avoid-laggy-interaction

**Reply by MohAid** [reply]
Hi @rodja, I'm sorry I used the word "frequently" but what I meant was in some cases the client may have a weak connection then he will face that (I might use @kleynjan approach). In the end, I was convinced that this problem must be solved by the client, and despite that, I will do the best I can as a developer as much as lovely NICEGUI enable me.

@falkoschindler I tried to use `reconnect_timeout` but I'm not sure why it did not work with me it did not extend the time before reloading or before showing the "Connection lost. Trying to reconnect" message.

**Comment by kleynjan** [comment]
@MohAid see code below for a rough example. This uses session state (app.storage.user) to keep "form" state across reloads and websocket reconnects. New invoices open in separate tabs (and can be edited there) per your spec. As Falko explained, we need a separate url for every invoice, so the pattern is /invoice/{invoice_id}

Note that I wouldn't recommend this approach (with multiple browser tabs) for a serious app: too much issues with controlling the lifecycle and focus across the tabs. Case in point, if you want to try out this code, use Chrome, as Safari doesn't allow opening tabs this way.

IMHO, if you want to have multiple invoices open for adding/editing, a better approach would be to use a single page with Nicegui tabs or dialogs. If you maintain state through app.storage.user, the user can still reload the entire page without losing information. Hope this helps! 

```python
from nicegui import app, ui
from fastapi import HTTPException

@ui.refreshable
def invoice_list(invs: list) -> None:
    ui.table(
        rows=[inv for inv in invs if inv["saved"]],
        columns=[{"name": col, "label": col, "field": col} for col in ["id", "customer", "amount"]],
    )

@ui.page("/")
def invoices() -> None:
    def open_new_invoice() -> None:
        invs = app.storage.user["invs"]
        if invs:
            next_id = max([inv["id"] for inv in invs]) + 1
        else:
            next_id = 1
        invs.append({"id": next_id, "customer": "", "amount": 0, "saved": False})
        ui.open(f"/invoice/{next_id}", new_tab=True)

    def clear_state() -> None:
        app.storage.user.clear()
        invoice_list.refresh()
        ui.open("/")

    if "invs" not in app.storage.user:
        app.storage.user["invs"] = [
            # {"id": 1, "customer": "Alice", "amount": 100, "saved": True},
            # {"id": 2, "customer": "Bob", "amount": 200, "saved": True},
        ]
 
    invoice_list(app.storage.user["invs"])
    ui.button("New invoice", on_click=open_new_invoice)
    ui.button("Clear session", on_click=clear_state)

@ui.page("/invoice/{id}")
def invoice(id: int) -> None:
    def save_invoice(inv: dict):
        inv["saved"] = True
        # no simple way to close a tab from the server side, so...
        ui.notify("Invoice saved. You may now close this tab... ")
        invoice_list.refresh()

    # get invoices for this user from session state and find inv by id
    try:
        invs = app.storage.user.get("invs", [])
        inv = [inv for inv in invs if inv["id"] == id][0]
    except IndexError as e:
        raise HTTPException(status_code=404, detail="Item not found") from e

    with ui.card():
        ui.label(f"Invoice # {id}")
        ui.input(label="Customer").bind_value(inv, "customer")
        ui.input(label="Amount").bind_value(inv, "amount")
    ui.button("Save", on_click=lambda: save_invoice(inv))

ui.run(storage_secret="abracadabra")
```

**Reply by MohAid** [reply]
@kleynjan Certainly, it helps. I appreciate your suggestions which opened the way for me to think about this issue, and the most important is they pushed me to continue with NICEGUI which I don't want to give up. Until now I have not reached this part, I still working on the basic pages to manage the users, when I reach this I'll see and come back here to tell you what approach I used, just to know. again I thank you so much for helping.

**Comment by rodja** [comment]
Because it is somehow related to this discussion: We are currently working on two more storage types: [`app.storage.tab`](https://github.com/zauberzeug/nicegui/pull/2837) which provides a place to store data across reloads and page changes in a single tab and [`app.storage.client`](https://github.com/zauberzeug/nicegui/pull/2820) which holds data only for the length of the current client connection (eg. will drop on every page change).
